pre/post-expand and post-jump actions

seletskiy opened the pull request:
Hi! It's me again.

I'm glad to introduce my shiny new PR, that introduces mechanism called "snippet actions". It makes possible to call arbitrary python code in certain snippet lifetime stages, like pre-expand, post-expand and post-jump.

I'd hope that help text is self-descriptive, so I will not repeat functionality description there.

**Note**: I've changed semantics of `cursor` value which is passed to previous mechanism "Context Snippets". After using it for a while I've found it highly inconvinient to use non-zero-based indexing for line numbers. I think there are not so many people who are use context snippets right now, so my change should not be devastating.

@SirVer: let's merge it fast. I have another productivity booster idea.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/sirver/ultisnips/507)
<!-- Reviewable:end -->

SirVer commented:
The python3 tests are broken - please look into that. 

Otherwise the feature looks very powerful and your design should work that way. I am concerned that it adds a ton of more complexity to UltiSnips that might not be appreciated by most users, but it certainly is powerful. 

I did start reviewing some of the code - not through it yet. But here is the first bunch of comments.

---

Review status: 3 of 16 files reviewed, 10 unresolved discussions, some commit checks failed.
Reviewed files:
- doc/UltiSnips.txt @ r1
- pythonx/UltiSnips/buffer_helper.py @ r1
- pythonx/UltiSnips/position.py @ r1

---

<sup>**[doc/UltiSnips.txt, line 1434 \[r1\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-Jqkb8FngKYjag5ZgqOO)** ([raw file](https://github.com/sirver/ultisnips/blob/d297395f80c1ff8f45f17f5cfdbde1226dccd15f/doc/UltiSnips.txt#L1434)):</sup>
is visual_content the same as snip.v inside the snippet? If so, I would mention it here. If not, I would suggest changing it so that it is.

---

<sup>**[doc/UltiSnips.txt, line 1510 \[r1\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-JqkbarnUtatkP6mSKwg)** ([raw file](https://github.com/sirver/ultisnips/blob/d297395f80c1ff8f45f17f5cfdbde1226dccd15f/doc/UltiSnips.txt#L1510)):</sup>
Why not sum up the variables and put them into an object, something like `snip`. Less polluting of namespaces and easier to extend in the future without breaking users. 

---

<sup>**[pythonx/UltiSnips/buffer_helper.py, line 7 \[r1\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-JqkZyt4N_NRrU_o-yit)** ([raw file](https://github.com/sirver/ultisnips/blob/d297395f80c1ff8f45f17f5cfdbde1226dccd15f/pythonx/UltiSnips/buffer_helper.py#L7)):</sup>
needs to derive from object to be a proper python2 object

---

<sup>**[pythonx/UltiSnips/buffer_helper.py, line 7 \[r1\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-Jqk_5P3Zi1gR2uP-qmU)** ([raw file](https://github.com/sirver/ultisnips/blob/d297395f80c1ff8f45f17f5cfdbde1226dccd15f/pythonx/UltiSnips/buffer_helper.py#L7)):</sup>
Add some class documentation. Helper is a very indistinct demoniker which makes it hard to understand what the class is supposed to do 

---

<sup>**[pythonx/UltiSnips/buffer_helper.py, line 11 \[r1\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-Jqk_FiYmePTq_6BoW3L)** ([raw file](https://github.com/sirver/ultisnips/blob/d297395f80c1ff8f45f17f5cfdbde1226dccd15f/pythonx/UltiSnips/buffer_helper.py#L11)):</sup>
What if somebody edits a 10gb file? Taking the full buffer copy should be avoided if possible.

---

<sup>**[pythonx/UltiSnips/buffer_helper.py, line 17 \[r1\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-Jqk_PvUt_7hgpwSDU13)** ([raw file](https://github.com/sirver/ultisnips/blob/d297395f80c1ff8f45f17f5cfdbde1226dccd15f/pythonx/UltiSnips/buffer_helper.py#L17)):</sup>
enumerate(self._buffer_copy)

---

<sup>**[pythonx/UltiSnips/buffer_helper.py, line 25 \[r1\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-Jqk_W9UJBBtEp6BFg6A)** ([raw file](https://github.com/sirver/ultisnips/blob/d297395f80c1ff8f45f17f5cfdbde1226dccd15f/pythonx/UltiSnips/buffer_helper.py#L25)):</sup>
it would be nice to actually handle such changes, would it not? Do you have any ideas for that?

---

<sup>**[pythonx/UltiSnips/buffer_helper.py, line 27 \[r1\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-Jqk_bBdcYOyA4fis88W)** ([raw file](https://github.com/sirver/ultisnips/blob/d297395f80c1ff8f45f17f5cfdbde1226dccd15f/pythonx/UltiSnips/buffer_helper.py#L27)):</sup>
might want to throw in a vim `:help` topic here for further reference.

---

<sup>**[pythonx/UltiSnips/buffer_helper.py, line 35 \[r1\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-Jqk_qBIKX0Z8haRvURt)** ([raw file](https://github.com/sirver/ultisnips/blob/d297395f80c1ff8f45f17f5cfdbde1226dccd15f/pythonx/UltiSnips/buffer_helper.py#L35)):</sup>
strip os.linesep

---

<sup>**[pythonx/UltiSnips/position.py, line 69 \[r1\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-JqkaQge0WG74Qm6S_rC)** ([raw file](https://github.com/sirver/ultisnips/blob/d297395f80c1ff8f45f17f5cfdbde1226dccd15f/pythonx/UltiSnips/position.py#L69)):</sup>
This reads super confusing at the callsite - please use .line and .col throughout and revert this again

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/sirver/ultisnips/507)
<!-- Sent from Reviewable.io -->

seletskiy commented:
@SirVer: Regarding unpopularity among users: I think we need to do something to promote right way of using snippet. README on GitHub page is lack of any advices.

Snippets is a very powerful instrument which can be used incorrectly and one will not gain any significant boost. Look at community snippets for python. Snippet for `if` is triggered by typing `if`. The most common statement is triggered by two characters. It's sad.

SirVer commented:
`if` for `if` seems like a logical trigger to me - remembering your triggers is the hardest part for me. I think contributing really powerful snippets to vim-snippets and/or regularly blogging about snippets would help a ton to make the more powerful features better known. What do you think?

seletskiy commented:
> is visual_content the same as snip.v inside the snippet? If so, I would mention it here. If not, I would suggest changing it so that it is.

Yes, it is same that `snip.v`. But `snip` object can not be used in that context, because other properties are just incorrect (like `snip.rv`). So choices there are: either use another `snip` object with single `v` property, or explicitly use `visual_selection`. Latest case is better, because it's not breaking expected and documented behaviour.

seletskiy commented:
> Why not sum up the variables and put them into an object, something like snip. Less polluting of namespaces and easier to extend in the future without breaking users.

Agree on that point. What do you suggest, to use `snip` name, but with different properties set?

seletskiy commented:
> What if somebody edits a 10gb file? Taking the full buffer copy should be avoided if possible.

Yeah, it will be an issue. That ugly solution was implemented for catching buffer changes which was made without buffer helper; I wonder how to track changes without buffer copy and see no way to do it (btw, just came to my mind, that `b:changetick` can be useful, I'll try it).

seletskiy commented:
> it would be nice to actually handle such changes, would it not? Do you have any ideas for that?
Unfortunately, I think there is no way to do it. Separate line insertions/deletions can be tracked via vim-marks (which I already use), but changes **in** the lines are not tracked by the marks and vim API lacks of any way to track it. I guess, that buffer proxy is the only solution. I thought about diffing entire buffer, but it is a) unreliable; b) ultra-slow.

seletskiy commented:
@SirVer: UltiSnips already has mechanism for dealing with variety of snippets, and this mechanism is interpolation.

I don't know how are you familiar with Go language, so there can three different type declarations:
```go
type MyTypeAlias AliasedType
```
```go
type MyStruct struct {
    Field FieldType
    ...
}
```
```go
type MyInterface interface {
    SomeFunc()
}
```

I've took a look at community snippets (honza/vim-snippets), and there is **no** snippet for `type`; `struct` (st) and `interface` (inf) snippets are **syntastically wrong**.

Obvious solution for case of three different type declarations is a three different snippets with different triggers. But, hey, there is an interpolation feature, and all that three snippets can be expanded via **one** trigger; UltiSnips will do the rest. Algorithm is simple:
* if user types `(` inside snippet body, then replace `struct` with `interface`;
* if user types `Blah`, it's type alias;
* if user types `Blah blah`, then it's a struct;

Look at my snippet: https://github.com/seletskiy/dotfiles/blob/8e04f6a47fa1509be96094e5c8923f4b49b22775/.vim/UltiSnips/go.snippets#L11-23

It covers all three cases using one single-character trigger. You don't need to remember three different snippets:

![snip](https://cloud.githubusercontent.com/assets/674812/7977989/d8b13138-0a7f-11e5-917f-8b1a536cac65.gif)

Regarding `if`: given that it's most popular construction, it should be `i` for almost all filetypes. Also, there are plugins like Unite and it's super-easy to open list of snippets for consult. `else` should not have specific tab-trigger at all (except for python, maybe); must be triggered by hitting `<TAB>` after enclosing `}` or by just hitting `<SPACE>` (I want to speak about that feature later).

And that without mention of context-snippets, which can cut the required amount of triggers even more. I even doubt that I can come with 44 (lower + upper case letters) common snippets for one filetype, keeping in mind interpolation+context.

SirVer commented:
Have you ever tried http://reviewable.io? It makes discussing so large code reviews much easier. 

> Agree on that point. What do you suggest, to use snip name, but with different properties set?

I think that is probably the most ergonomic way to go. It might be confusing that snip has different properties in different contexts, but taking more global variable space is also not great. 

> Unfortunately, I think there is no way to do it. 

Neovim sends information about changed lines through it's msgpack API. This is so that UIs know which lines need redrawing. I asked before if we could have the same for full buffers - which would be a possible solution. Alternatively we could add better mark support into Neovim. But I agree for vanilla Vim there is really no good solution :/. 
 
>  I thought about diffing entire buffer, but it is a) unreliable; b) ultra-slow.

I did that in earlier versions of UltiSnips :).

[regarding your snippet use]
I think your use of UltiSnips is exemplary. We should get your snippets into vim-snippets and we should write a short tutorial for how to use them so that we get better community contributions going forward.

SirVer commented:
A couple more comments. And you did not address all of the old ones. 

---

Review status: 15 of 16 files reviewed, 15 unresolved discussions, all commit checks successful.
Reviewed files:
- pythonx/UltiSnips/_vim.py @ r1
- pythonx/UltiSnips/buffer_helper.py @ r2
- pythonx/UltiSnips/snippet/definition/_base.py @ r1
- pythonx/UltiSnips/snippet/definition/snipmate.py @ r1
- pythonx/UltiSnips/snippet/source/file/_common.py @ r1
- pythonx/UltiSnips/snippet/source/file/ultisnips.py @ r1
- pythonx/UltiSnips/snippet_manager.py @ r1
- pythonx/UltiSnips/text_objects/_snippet_instance.py @ r1
- syntax/snippets.vim @ r1
- test/test_ContextSnippets.py @ r1
- test/test_Editing.py @ r1
- test/test_UltiSnipFunc.py @ r1
- test/vim_test_case.py @ r1

---

<sup>**[pythonx/UltiSnips/snippet/definition/_base.py, line 107 \[r2\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-JrGqsdHE0hsWc4zVJSw)** ([raw file](https://github.com/sirver/ultisnips/blob/578065c5b8241f42d24097971c96a6816540892f/pythonx/UltiSnips/snippet/definition/_base.py#L107)):</sup>
why did you change context to holder? holder sounds really foreign - what do you want to express here? Maybe container is a better word.

---

<sup>**[pythonx/UltiSnips/snippet/definition/_base.py, line 128 \[r2\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-JrGrEuzDSMkfaHaVEpP)** ([raw file](https://github.com/sirver/ultisnips/blob/578065c5b8241f42d24097971c96a6816540892f/pythonx/UltiSnips/snippet/definition/_base.py#L128)):</sup>
How about building a python context manager that restores a mark at it's end. That way it is impossible to forget resetting the mark.

---

<sup>**[pythonx/UltiSnips/snippet_manager.py, line 423 \[r2\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-JrGs5b1OSe7iXm1Lh1_)** ([raw file](https://github.com/sirver/ultisnips/blob/578065c5b8241f42d24097971c96a6816540892f/pythonx/UltiSnips/snippet_manager.py#L423)):</sup>
same comment about using a contextmanager

---

<sup>**[pythonx/UltiSnips/snippet_manager.py, line 460 \[r2\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-JrGsBYDQFjRHa_SeEle)** ([raw file](https://github.com/sirver/ultisnips/blob/578065c5b8241f42d24097971c96a6816540892f/pythonx/UltiSnips/snippet_manager.py#L460)):</sup>
thinking some more about this, why not just use "backwards" and "forwards" ?

---

<sup>**[test/vim_test_case.py, line 124 \[r2\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-JrGsjRz-RIuYEyAYdeD)** ([raw file](https://github.com/sirver/ultisnips/blob/578065c5b8241f42d24097971c96a6816540892f/test/vim_test_case.py#L124)):</sup>
what is that for?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/sirver/ultisnips/507)
<!-- Sent from Reviewable.io -->

kovetskiy commented on pythonx/UltiSnips/buffer_helper.py:
```diff
@@ -0,0 +1,101 @@
+    def validate_buffer(self):
+        if self.is_buffer_changed_outside():
+            raise RuntimeError('buffer was modified using vim.command or ' +
+            'vim.current.buffer; that changes are untrackable and leads to' +
```
I got this message, please fix this:
`...and leads toerrors in...`

kovetskiy commented:
What am I doing wrong?
```
global !p
def expand_new_snippet_on_first_jump():
	if tabstop == 1:
		from UltiSnips import UltiSnips_Manager
		UltiSnips_Manager.expand_anon("some_check($1)")
		return 'keep'
endglobal

post_jump "new_cursor = expand_new_snippet_on_first_jump()"
snippet "test" "test new features" "True" bwre
if $1:
	$2
endsnippet
```

I got:
```
An error occured. This is either a bug in UltiSnips or a bug in a
snippet definition. If you think this is a bug, please report it to
https://github.com/SirVer/ultisnips/issues/new.

Following is the full stack trace:
Traceback (most recent call last):
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/snippet_manager.py", line 56, in wrapper
    return func(self, *args, **kwds)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/snippet_manager.py", line 226, in expand_anon
    self._do_snippet(snip, before)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/snippet_manager.py", line 590, in _do_snippet
    self._csnippets[0].replay_user_edits(edit_actions)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_snippet_instance.py", line 53, in replay_user_edits
    self._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 223, in _do_edit
    child._do_edit(cmd, ctab)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 268, in _do_edit
    self._child_has_moved(idx, pivot, delta)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 286, in _child_has_moved
    self._parent._children.index(self), pivot, diff
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 282, in _child_has_moved
    child._move(pivot, diff)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 271, in _move
    TextObject._move(self, pivot, diff)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/text_objects/_base.py", line 70, in _move
    self._start.move(pivot, diff)
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/position.py", line 19, in move
    if self < pivot:
  File "/home/vxw/.vim/bundle/ultisnips/pythonx/UltiSnips/position.py", line 61, in __lt__
    return (self.line, self.col) < (other.line, other.col)
RuntimeError: maximum recursion depth exceeded in cmp
```

It can be fixed? post_jump actions is shared to anon snippets?

seletskiy commented:
@kovetskiy: try it now. Should be fixed.

seletskiy commented:
> This reads super confusing at the callsite - please use .line and .col throughout and revert this again.

It's made intentionally to keep interface same with `vim.current.window.cursor`. I have a plenty of library functions that are accept cursor and it's quite inconvinient to convert the value back and forth.

> Have you ever tried http://reviewable.io? It makes discussing so large code reviews much easier. 

@SirVer: I dislike it. It's very slow (much slower rendering and loading, than GitHub), and information presentation is ugly (what are that "broken paper" pictures for?)

> thinking some more about this, why not just use "backwards" and "forwards" ?

It's easier to access previous tabstop by using simple math op (`prev_tabstop = snip.tabstops[snip.tabstop - snip.jump_direction]`).

> I think that is probably the most ergonomic way to go. It might be confusing that snip has different properties in different contexts, but taking more global variable space is also not great. 

I've moved all state variable to different `snip` object. I need to refactor that solution, however, but now it's working prototype.

> Neovim sends information about changed lines through it's msgpack API. This is so that UIs know which lines need redrawing. I asked before if we could have the same for full buffers - which would be a possible solution. Alternatively we could add better mark support into Neovim. But I agree for vanilla Vim there is really no good solution :/. 

I'm not using neovim right now and will not use it until authors make native multiple cursors support. Till that I see no point of switching, so I propose to leave that integration for the future.

> What if somebody edits a 10gb file? Taking the full buffer copy should be avoided if possible.
I've tried to use `&changedtick` for that purpose and succeeded. It should not be a problem anymore.

@SirVer: I've also fixed two bugs that were in original UltiSnips branch. They are reflected by testcases `TabStop_KeepCorrectJumpListOnOverwriteOfPartOfSnippet` and `DoNotCrashOnUndoAndJumpInNestedSnippet`.

Regarding all other comments of yours: I agree and some of them already fixed.

SirVer commented:
Is this ready for another look?

seletskiy commented:
@SirVer: yep. Please, take a look.

seletskiy commented:
@SirVer: oh, I'm sick because of python3 vs. python2 quirks. I've moved code from `UltiSnips/__init__.py` to `snippet_manager`, my python3 setup for whatever reason loaded `__init__.py` twice, which caused duplication of the `UltiSnips_Manager` object (in the travis, however, there was only one failed test regarding that issue, don't know why). After moving the code all tests passing fine.

seletskiy commented:
@SirVer: ping?

SirVer commented:
Reviewed 5 of 11 files at r4, 3 of 4 files at r5, 8 of 8 files at r6.
Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

<sup>**[doc/UltiSnips.txt, line 1442 \[r1\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-Jqkb8FngKYjag5ZgqOO)** ([raw file](https://github.com/sirver/ultisnips/blob/d297395f80c1ff8f45f17f5cfdbde1226dccd15f/doc/UltiSnips.txt#L1442)):</sup>
Please check wrapping here - some text is longer than 80 columns

---

<sup>**[pythonx/UltiSnips/__init__.py, line 4 \[r6\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-JtmUS99XHR1c9riBzr7)** ([raw file](https://github.com/sirver/ultisnips/blob/671134e67f491aaf9423d9012f68b01735a9cb67/pythonx/UltiSnips/__init__.py#L4)):</sup>
Please revert this file - there are other plugins relying on these definitions 

---

<sup>**[pythonx/UltiSnips/snippet/definition/_base.py, line 15 \[r6\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-JtmV8PX_txaGIbQwaD5)** ([raw file](https://github.com/sirver/ultisnips/blob/671134e67f491aaf9423d9012f68b01735a9cb67/pythonx/UltiSnips/snippet/definition/_base.py#L15)):</sup>
sort imports alphabetically

---

<sup>**[pythonx/UltiSnips/snippet/definition/_base.py, line 108 \[r2\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-JrGqsdHE0hsWc4zVJSw)** ([raw file](https://github.com/sirver/ultisnips/blob/578065c5b8241f42d24097971c96a6816540892f/pythonx/UltiSnips/snippet/definition/_base.py#L108)):</sup>
I still do not get 'holder' as a name. Care to explain?

---

<sup>**[pythonx/UltiSnips/text_objects/_base.py, line 181 \[r6\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-JtmVcS-m8L_QZKjMEoR)** ([raw file](https://github.com/sirver/ultisnips/blob/671134e67f491aaf9423d9012f68b01735a9cb67/pythonx/UltiSnips/text_objects/_base.py#L181)):</sup>
Wait... what? Does that work? should that not be 

`(a == b) and (b ==c)`?

---

<sup>**[pythonx/UltiSnips/text_objects/_python_code.py, line 51 \[r6\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-JtmW3OtE-TSQ6H23zxO)** ([raw file](https://github.com/sirver/ultisnips/blob/671134e67f491aaf9423d9012f68b01735a9cb67/pythonx/UltiSnips/text_objects/_python_code.py#L51)):</sup>
cache? save_cursor_pos? preserve sounds like an endangered animal.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/sirver/ultisnips/507)
<!-- Sent from Reviewable.io -->

SirVer commented:
Also, please rebase or merge master. The test suite is now better - i.e. it now actually tests on different python versions.

seletskiy commented:
> Please revert this file - there are other plugins relying on these definitions 

There is a problem with that import in python3. I've tried to debug one failed test from travis and noticed, that `__init__.py` file executed twice (it was literally that way, I've inserted `print` statements and seen two results). So, `UltiSnips_Manager` object got instantiated twice under python3. It became critical in that PR because `UltiSnips_Manager` can be called directly from snippet, and when snippet gets stale object things became screwed up.

I guess the reason of that behaviour is very similar to reasons described in the article http://python-notes.curiousefficiency.org/en/latest/python_concepts/import_traps.html.

Reverting is not a solution (it will break functionality and tests) and do not know how to fix it.

SirVer commented:
Have you tried having a guardian variable? Similar to a pragma once? Moving the variable out of the top level namespace will break all python plugins that use UltiSnips. Since I do not know how many there are, I'd rather not do that.

SirVer commented:
I also changed how importing works in master now - maybe it is no longer importing twice now.

seletskiy commented:
> Wait... what? Does that work? should that not be
> (a == b) and (b ==c)?

Yes, it works.

seletskiy commented:
>  cache? save_cursor_pos? preserve sounds like an endangered animal.

Because it is. Without this call snippet will not work and things can be messed up. `cache` do not reflect meaning and `save_cursor_pos` looks bad in call `snip.cursor.save_cursor_pos`.

SirVer commented:
Reviewed 7 of 7 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

<sup>**[doc/UltiSnips.txt, line 1442 \[r1\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-Jqkb8FngKYjag5ZgqOO)** ([raw file](https://github.com/sirver/ultisnips/blob/d297395f80c1ff8f45f17f5cfdbde1226dccd15f/doc/UltiSnips.txt#L1442)):</sup>
wrapping is still messed up.

---

<sup>**[pythonx/UltiSnips/__init__.py, line 4 \[r6\]](https://reviewable.io:443/reviews/sirver/ultisnips/507#-JtmUS99XHR1c9riBzr7)** ([raw file](https://github.com/sirver/ultisnips/blob/671134e67f491aaf9423d9012f68b01735a9cb67/pythonx/UltiSnips/__init__.py#L4)):</sup>
you only partially reverted? did it not work?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/sirver/ultisnips/507)
<!-- Sent from Reviewable.io -->

SirVer commented:
>> Wait... what? Does that work? should that not be
>> (a == b) and (b ==c)?

> Yes, it works.

It is very unreadable. Could you rewrite to my suggestions please?

> snip.cursor.save_cursor_pos.

I agree, but what about `snip.cursor.[save|remember|cache]_position`?

Also, tests are broken on Neovim and a bunch of my older comments are not resolved. maybe this was not ready for another look yet? If so, sorry that I interjected.

seletskiy commented:
@SirVer: I believe I've fixed everything right now. If I missed something, pls, point it.

> It is very unreadable. Could you rewrite to my suggestions please?

In my opinion it's not better nor worse than code two lines above: `if children._start <= pos < children._end:`. Anyway, I've changed it to your proposal.

> I agree, but what about snip.cursor.[save|remember|cache]_position?

You use same word in file pythonx/UltiSnips/vim_state.py (`class VisualContentPreserver`). In my opinion `snip.cursor.preserve()` reflects best what happens with cursor state inside the UltiSnips engine, all other names either ugly or do not reflect the meaning.


>  you only partially reverted? did it not work? [regarding __init__.py]

In the original `__init__.py` file `UltiSnips_Manager` is only exported name, so I believe other plugins should work fine. I tested it with YouCompleteMe and it's working fine (I can see snippets inside completion menu).

I've tried to use guardian variables with no success. It was like python3 just evaluated file twice _without_ keeping variable state from _single_ import statement (yeah, sounds like total nonsense, but I saw it even when inserting print statements through all the code).

Try to insert `print('test')` on the first non-empty line in `pythonx/UltiSnips/__init__.py` and run vim with python3 support. See how many times `test` will be printed to `:mess`. In my setup it prints twice while using vim-python3 (total WTF) and only once using vim-python2.

seletskiy commented:
@SirVer: Can we merge it already? Little bit tired of resolving merge conflicts.

SirVer commented:
Reviewed 5 of 6 files at r9, 3 of 3 files at r10.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/sirver/ultisnips/507)
<!-- Sent from Reviewable.io -->

SirVer commented:
Merged :) Thanks Stanislav!

kovetskiy commented:
Awesome! Very cool stuff!
